Implement System.Diagnostics.Process on Linux

stephentoub opened the pull request:
This is an initial implementation of the System.Diagnostics.Process
assembly for Linux.  There are some rough edges here: basic process
enumeration and inspection should work, as should waiting for processes
through various means (WaitForExit, Exited event, etc.), but other
functionality isn't ready for primetime yet, namely Start, which has been
roughly implemented but which hasn't been tested yet.

Some of the surface area doesn't have good equivalents on Linux.  Where I
thought it reasonable, in some cases I just left default values in place
or ignore things. In other cases, I throw PlatformNotSupportedException.
We can tweak this over time.

The implementation gets most of its data from the procfs file system.
Suppot for waiting is the most complicated area of the implementation, and
probably still has a few bugs to sort out.

stephentoub commented:
cc: @ellismg, @pgavlin, @jkotas

ellismg commented:
Setting expectations Steve, I won't get to take a careful look at this till tomorrow.

stephentoub commented:
Thanks, Matt. Sounds good. I need to fix something small that I broke on the Windows side, anyway.

pgavlin commented:
Same here: probably won't have eyes on this until tomorrow.

stephentoub commented:
Thanks, Maxwe11. In this particular case I chose not to do that as Process.dll doesn't already include StringBuilderCache.cs, plus the relevant savings from that would be fairly small given the other costs involved here and the relative infrequency with which new processes are launched. If I was concerned about allocations in this particular section of code, there would be  other likely targets for improvement :) Still, I'll keep it in mind, and if we have other reasons to bring in a cache to this assembly, we should certainly then use it here.

stephentoub commented:
@dotnet-bot test this please

(Build failed due to #659.  I submitted a PR to disable the failing test for now in #660.)

stephentoub commented:
@dotnet-bot test this please

aleks-sidorenko commented:
Quick question: do you have convention on naming partial classes for different platforms? I mean sometimes I see for Windows: Class.Windows.partial.cs or Class.Win32.partial.cs, and for Unix based systems: Class.Unix.partial.cs or Class.Linux.partial.cs.
Do you have single convention?

stephentoub commented:
@aleks-sidorenko, this is the convention we're currently following:
https://github.com/dotnet/corefx/wiki/Cross-Platform-Code-Guidelines#how-should-the-platform-specific-files-be-named-eg-filestreamwindowscs-win32filestreamcs

ellismg commented on src/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.cs:
```diff
@@ -0,0 +1,321 @@
+
+            int index = 0;
+            int i = 0;
+            for (; i < ExpectedStatFileParts && index < statFileContents.Length; i++)
```
I understand the desire to save an allocation here, which leaves the code structured the way it is.   My first question is what happens in the case where we early out of the loop because index ends up being greater than ```statFileContents.Length``` before we have parsed all the parts?  In Debug builds, we would assert, but in release builds we'll just return a ParsedStat structure that is incomplete.  Is that what we actually want?

Would it be possible to structure the code in the following way?  Create a mutable struct that is a cursor into the string.  Have that struct have methods like GetStringOrFail() GetInt32OrFail() GetInt64OrFail() which advance the cursor and return a value.

Then we could rid ourselves of this loop entirety and just have straight line code like:

```
results.flags = cursor.GetUInt32OrFail();
results.majflt = cursor.GetUInt64OrFail();
```

The methods could then either throw an exception in the case where the input string didn't have enough parts or just Debug.Assert() and return dummy values if we want to maintain the partial parsing behavior (which seems dubious to me long term, but who knows).

Thoughts?

stephentoub replied:
> Would it be possible to structure the code in the following way?

Excellent suggestion; I've done that and it not only helps clean up the code, it makes it easier to skip over creating strings for fields we're not actually using anywhere.

> In Debug builds, we would assert, but in release builds we'll just return a ParsedStat structure that is incomplete. Is that what we actually want?

I went back and forth on this.  Since you obviously lean towards making it exceptional, that's what I've now done.

ellismg commented on src/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.cs:
```diff
@@ -0,0 +1,321 @@
+
+        private static string GetStatFilePathForProcess(int pid)
+        {
+            return RootPath + pid.ToString(CultureInfo.InvariantCulture) + "/stat";
```
Do you want to constify "/stat" since you use it in a few places?  At that point maybe you do "/task/" as well?

ellismg commented on src/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.cs:
```diff
@@ -0,0 +1,321 @@
+                        }
+                        else
+                        {
+                            Debug.Fail("Unexpected comm format");
```
What's the rationale for not failing in release here?  We have this problem all over the place.  Do we expect up stack stuff to be tolerant to garbage returned by this function?

Note we do this in the parsing logic later as well.

stephentoub replied:
Per my comment above, I've changed all of this to throw.

ellismg commented on src/System.Diagnostics.Process/src/Microsoft/Win32/SafeHandles/SafeThreadHandle.Unix.cs:
```diff
@@ -19,17 +18,18 @@ namespace Microsoft.Win32.SafeHandles
         protected override bool ReleaseHandle()
         {
-            throw NotImplemented.ByDesign; // TODO: Implement this
+            // Nop.
```
I assume this is due to the same rationale as SafeProcessHandle.  Do you mind including that here?

stephentoub replied:
Actually, it turns out that SafeThreadHandle isn't used at all in the Unix implementation.  I've recombined the Windows partial types, deleted the Unix one, and removed it from the Unix build entirely.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/ModuleInfo.cs:
```diff
@@ -12,10 +12,12 @@ namespace System.Diagnostics
     /// <internalonly/>
     internal sealed class ModuleInfo
     {
+#pragma warning disable 0649 // unused fields on some platforms
```
Thoughts on PALifying this instead?  Not worth it to just remove the pragmas?

stephentoub replied:
I've done it this way because I think some of this could potentially be supported at some point, I'm just not sure how to do it at present.

ellismg replied:
Cool.  Sounds good.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs:
```diff
@@ -0,0 +1,589 @@
+        private bool PriorityBoostEnabledCore
+        {
+            get { return false; } //Nop
+            set { } // Nop
```
Thoughts on throwing here?

stephentoub replied:
I think the getter should return false.  I was 50/50 on the setter; since you seem to prefer the exception, I've changed it to throw PlatformNotSupportedException.

ellismg replied:
I agree with your stance on the getter.

stephentoub commented:
Thanks, @ellismg!  I believe I've addressed all of your feedback.  I also made an additional perf tweak, which was necessary for testing some things due to the current GC limitations.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs:
```diff
@@ -0,0 +1,696 @@
+        public void Kill()
+        {
+            EnsureState(State.HaveId);
+            int errno = Interop.libc.kill(_processId, Interop.libc.Signals.SIGKILL);
```
Do you have an feeling for if TerminateProcess (which is what we do on Windows) maps more to SIGKILL vs SIGTERM?

stephentoub replied:
Processes can ignore SIGTERM; they can't ignore SIGKILL.  Since a process can't ignore TerminateProcess either, I went with SIGKILL.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs:
```diff
@@ -0,0 +1,696 @@
+        /// </summary>
+        private ProcessPriorityClass PriorityClassCore
+        {
+            // This mapping is relatively arbitrary.  0 is normal based on the man page,
```
Might be nice to document this somewhere (once we have a docs story) and make the mapping contractual.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs:
```diff
@@ -0,0 +1,696 @@
+                // If we're redirecting streams, configure stdin, stdout, stderr
+                if (stdinPipeFds != null)
+                {
+                    Redirect(stdinPipeFds[0], Interop.libc.FileDescriptors.STDIN_FILENO);
```
Consider having consts in this method for ReadEndOfPipe and WriteEndOfPipe to make the indexing more obvious?

stephentoub replied:
Good idea, I'll do that.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs:
```diff
@@ -0,0 +1,696 @@
+                ParseArgumentsIntoList(startInfo.Arguments, argv);
+                
+                // Parse the environment variables
+                var envVars = new List<string>(startInfo.Environment.Count);
```
Why not just create an array here, since you know the final size?

stephentoub replied:
Yup, silly, will fix.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs:
```diff
@@ -0,0 +1,696 @@
+
+            // Then check each directory listed in the PATH environment variables
+            string pathEnvVar = Environment.GetEnvironmentVariable("PATH");
+            if (pathEnvVar != null)
```
Is it possible to use the mutable string walker thing from earlier here as well?

stephentoub replied:
Yup, I've separated it out into a reusable helper in Common\src\System\IO\StringParser.cs.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs:
```diff
@@ -0,0 +1,696 @@
+        internal static DateTime BootTimeToDateTime(ulong ticksAfterBoot)
+        {
+            // Read procfs to determine the system's uptime, aka how long ago it booted
+            string uptimeStr = File.ReadAllText(Interop.procfs.ProcUptimeFilePath);
```
I'm sure the sniffing would do the right thing, but can you pass an explicit encoding here (UTF8, I assume?)

stephentoub replied:
Done.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs:
```diff
@@ -0,0 +1,503 @@
+                Debug.Assert(foundState);
+                if (foundState)
+                {
+                    if (--pws._outstandingRefCount == 0)
```
Would you mind doing the decrement outside the test?  I know they are isomorphic, but I think it reads better.

stephentoub replied:
Done.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs:
```diff
@@ -0,0 +1,503 @@
+    //   share the same ProcessWaitState.  And since they are sharing the same wait state object,
+    //   the wait state object uses its own lock to protect the per-process state.  This includes
+    //   caching exit / exit code / exit time information so that a Process object for a process that's already
+    //   had waitpid called for it can get at its exit information.
```
So it seems like the one problem with this approach is that if a process exits, but there are outstanding wait handles handed out (and rooted, so they can't be GC'd), then a new process is created and the pid is recycled, new calls to get that process's wait state will get the old processe's WaitState.  Is that correct?

I'm not really sure what we can do here, but we might as well call it out and say we understand this problem and there isn't a mitigation?

stephentoub replied:
That's correct. I'll add a comment to that effect.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs:
```diff
@@ -0,0 +1,503 @@
+
+                        // PERF NOTE:
+                        // At the moment, we never call CheckForExit with blockingAllowed:true (which in turn allows
+                        // waitpid to block until the child has completed) because we currently allows call it while
```
I think "because we currently allow call it" should be "because we currently call it"

stephentoub replied:
Done.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs:
```diff
@@ -0,0 +1,503 @@
+            }
+        }
+
+        private void CheckForExit(bool blockingAllowed = false)
```
Would you mind making a comment here about how we never call this with blockingAllowed = true and reference the comment below?  Alternately could you update the comment to have the string CheckForExit(true) in it, so CTRL+F finds it.

From the comment later in the file it was not clear to me if it's safe to call this with "true" today (outside of the _gate lock) or we would need to do additional auditing wrt torn reads.

Do you expect to make this optimization soon?  I would sort of just prefer that we not have this code in here, especially if we are not confidant of it yet.

stephentoub replied:
I'd like to leave this here as a reminder that it's really something we should do, or should at least investigate to see what kind of impact it has on perf (I expect it will be a valuable thing to do).  For now, I've added the suggested comments as well as a Debug.Assert that validates it's not set to true until we do more work.

ellismg commented on src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs:
```diff
@@ -0,0 +1,503 @@
+
+        private void CheckForExit(bool blockingAllowed = false)
+        {
+            Debug.Assert(Monitor.IsEntered(_gate));
```
I would assume you would want to also assert that if Monitor.IsEntered(_gate) then blockingAllowed == false?

stephentoub replied:
Done.

ellismg commented:
I made a few more comments.  Overall, this looks good to me.  I am excited to get to the point where we can actually start testing more of this.

Sorry for taking so long to slog through all of this.

stephentoub commented:
Thanks for all the good feedback, Matt.  I've addressed your comments and updated the PR with the changes.

ellismg commented:
Thanks Steve.  LGTM.

stephentoub commented:
Thanks, Matt.
